libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -2150,7 +2152,8 @@ uint16_t AP_Mission::get_landing_sequence_start()
                 // command does not have a valid location and cannot get next valid
                 continue;
             }
-            float tmp_distance = tmp.content.location.get_distance(current_loc);
+            tmp.content.location.change_alt_frame(Location::AltFrame::ABSOLUTE);
```

Comment:
If `get_distance_NED` needs to have both locations in the same altitude frame then it is simply busted and should be fixed.

Reply:
> If `get_distance_NED` needs to have both locations in the same altitude frame then it is simply busted and should be fixed.

well, the note above the function says
 ```// return the distance in meters in North/East/Down plane as a N/E/D vector to loc2, NOT CONSIDERING ALT FRAME!```
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Common/Location.cpp#L268-L274

Reply:
@peterbarker What do you think about this? I haven't tested it but seems pretty straightforward. I could make a PR for it..

```
Vector3f Location::get_distance_NED(const Location &loc2) const
{
    int32_t alt1_cm = alt;
    int32_t alt2_cm = loc2.alt;

    if (get_alt_frame() != loc2.get_alt_frame()) {
        get_alt_cm(AltFrame::ABSOLUTE, alt1_cm);
        loc2.get_alt_cm(AltFrame::ABSOLUTE, alt2_cm);
    }

    return Vector3f((loc2.lat - lat) * LOCATION_SCALING_FACTOR,
                    diff_longitude(loc2.lng,lng) * LOCATION_SCALING_FACTOR * longitude_scale((lat+loc2.lat)/2),
                    (alt1_cm - alt2_cm) * 0.01);
}

Reply:
Why not skip the comparison of the frames and just get altitude straight from the first, and ask for the altitude in the second's frame? That has a fast path that checks if frame == current frame, and limits you to never having more then one conversion per call.

Reply:
@WickedShell because get_alt_cm() can return false if you're asking for the frame in a non-absolute frame

Reply:
> @peterbarker What do you think about this? I haven't tested it but seems pretty straightforward. I could make a PR for it..
> 
> ```
> Vector3f Location::get_distance_NED(const Location &loc2) const
> {
>     int32_t alt1_cm = alt;
>     int32_t alt2_cm = loc2.alt;
> 
>     if (get_alt_frame() != loc2.get_alt_frame()) {
>         get_alt_cm(AltFrame::ABSOLUTE, alt1_cm);
>         loc2.get_alt_cm(AltFrame::ABSOLUTE, alt2_cm);
>     }
> 
>     return Vector3f((loc2.lat - lat) * LOCATION_SCALING_FACTOR,
>                     diff_longitude(loc2.lng,lng) * LOCATION_SCALING_FACTOR * longitude_scale((lat+loc2.lat)/2),
>                     (alt1_cm - alt2_cm) * 0.01);
> }
> ```

Problem is the current signature doesn't allow for failure, and `get_alt_cm` can definitely fail.

Perhaps we need a new method `bool get_offset_NED(Vector3f &ret)` ?  If you think that's too much work we can do it later.

Reply:
I've removed this whole scope creep distraction by only doing a 3D distance check if alt frames can match, else do the existing 2D.